presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,224 @@
+            Supplier<TransactionalMetadata> hiveMetadataFactory,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "sync_partition_metadata",
+                ImmutableList.of(
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("mode", VARCHAR)),
+                SYNC_PARTITION_METADATA.bindTo(this));
+    }
+
+    public void sync_partition_metadata(ConnectorSession session, String schemaName, String tableName, String mode)
+    {
+        SyncMode syncMode = toSyncMode(mode);
+        HdfsEnvironment.HdfsContext context = new HdfsEnvironment.HdfsContext(session, schemaName, tableName);
+        SemiTransactionalHiveMetastore metastore = ((HiveMetadata) hiveMetadataFactory.get()).getMetastore();
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+
+        Table table = metastore.getTable(schemaName, tableName).orElseThrow(() -> new TableNotFoundException(schemaTableName));
+        checkArgument(!table.getPartitionColumns().isEmpty(), format("Table %s.%s is not partitioned", schemaName, tableName));
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElseThrow(() -> new TableNotFoundException(schemaTableName));
+            List<String> partitionsInFileSystem = listDirectory(fileSystem, fileSystem.getFileStatus(defaultLocation), table.getPartitionColumns(), table.getPartitionColumns().size()).stream()
+                    .map(fileStatus -> defaultLocation.toUri().relativize(fileStatus.getPath().toUri()).getPath())
+                    .collect(toImmutableList());
+
+            // partitions in metastore but not in file system
+            Set<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            // partitions in file system but not in metastore
+            Set<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+
+            if (syncMode == SyncMode.ADD || syncMode == SyncMode.FULL) {
+                addPartitions(metastore, session, table, defaultLocation, partitionsToAdd);
+            }
+            if (syncMode == SyncMode.DROP || syncMode == SyncMode.FULL) {
+                dropPartitions(metastore, session, table, partitionsToDrop);
+            }
+            metastore.commit();
+        }
+        catch (IOException e) {
```

Comment:
Add rollback. Eg we should rollback when addPartitions succeeds but dropPartitions fails.

Reply:
The `addPartition` and `dropPartition` don't throw `IOException` 
 exception. So that they will buffer all partitions to add and drop and commit them in one batch in `commit()` phase. From what I understand, the way `SemiTransactionalHiveMetastore` is implemented is similar to optimistic concurrency control. It calls `setShared()` and add partition actions to a local map in `addPartition` and `dropPartition`. During commit phase it will apply all `partitionActions` in `commitShared()`.

Reply:
I think the correct thing to do here is to move this `catch (IOException e)` so that it is only around the filesystem operation.  This will make is clear what can fail.  

The `addPartition` and `dropPartition` methods don't throw, and `commit` will do a best effort rollback on a partial failure.  If you call `rollback` after `commit` is called you will get an exception.